packages/core/upload/index.d.ts

```diff
@@ -0,0 +1,23 @@
+import type { ReadStream } from 'fs';
+
+declare module '@strapi/plugin-upload' {
+  export interface File {
+    name: string;
+    alternativeText?: string;
+    caption?: string;
+    width?: number;
+    height?: number;
+    formats?: Record<string, unknown>;
+    hash: string;
+    ext?: string;
+    mime: string;
+    size: number;
+    url: string;
+    previewUrl?: string;
+    path?: string;
+    provider?: string;
+    provider_metadata?: Record<string, unknown>;
+    stream?: ReadStream;
+    buffer?: Buffer;
+  }
+}
```

Comment:
This too. Although I don't think it's the end of the world. You have `MouseEvent` in both `React` and `DOM`.

Reply:
I would agree, then in this case it probably needs to be clearer i.e. this isn't the `FileEntity` but we could make that and then extend `IFile` which is the server object that extends `FileEntity` and adds the methods. I think this is a very good opportunity to define the "rules" for this system imo.

Reply:
agree, this could be a more general discussion for V5 and how and where we share types. Normally I would say, if front and back need to share types, they should be in the plugin root path like this index.d.ts. So I would agree this interface could be the one shared from front & back.

Reply:
This is temporary and will never be kept :) We will want to get that split into a generate type coming from the content types itself and a dedicated type for the uploading (with stream/buffer properties) 

I'm just moving code to ts so we can avoid v4-v5 conflicts.

Reply:
@Marc-Roig  & @Convly would love to get your opinion considering the context, unless there is a third option I'm not seeing I'll probably duplicate and make types as "small" as they need to be per provider.

Reply:
Types sharing, generation & exposure (including between admin/server) are topics that we're going to cover in the next TS support iteration before the V5.
If this is temporary & necessary to migrate the providers, then I don't mind if we go ahead with this (it's gonna be replaced anyway).
Adding "internal" flags to internal stuff that can be broken at any time should be a practice we enforce (we don't have any convention at the moment).
For instance, you could add `/** @internal */` above the type you want to keep away from users.

Reply:
So yeah, either duplication the type in each provider package or adding an internal jsdoc flag. (or both :sweat_smile: )

Reply:
agree with @Convly , both ways would work for me too.

I normally used the /** @internal */ tag and set `stripInternal` in the tsconfig, so it is not added in the generated .d.ts file

Reply:
@Marc-Roig Yes when you transpile it will avoid dumping them in the declaration file sadly here we manually create the declaration file so it's not going to do anything. and at worse since we use it as a public interface it would prevent me from using it if it was fully supported so defeating the prupose. I duplcated the interfaces to avoid this for now. Once we have everything in ts we can start connecting the dots 👍